crates/net/network/src/transactions/mod.rs

```diff
@@ -1015,25 +1015,46 @@ where
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         let this = self.get_mut();
 
+        // If the budget is exhausted we manually yield back control to tokio. See
+        // `NetworkManager` for more context on the design pattern.
+        let budget = &mut 1024;
+        macro_rules! budget_checkpoint {
+            () => {
+                *budget -= 1;
+                if *budget == 0 {
+                    // make sure we're woken up again
+                    return Poll::Ready(())
+                }
+            };
+        }
+
         // drain network/peer related events
         while let Poll::Ready(Some(event)) = this.network_events.poll_next_unpin(cx) {
+            budget_checkpoint!();
```

Comment:
I realize that this is problematic with the existing while loops because this can starve the downstream while loops if we return pending early

so I think we should reconsider how we poll the underlying streams